shell/browser/notifications/mac/cocoa_notification.mm

```diff
@@ -60,68 +66,160 @@
-      }
-    }
-    i++;
+    // SAP-15908 - Refine notification-activation for cold start in macOS
+    if (action.type == electron::NotificationAction::sTYPE_TEXT)
+      replyKey_ = [[NSString alloc]
+          initWithBytes:action.arg.data()
+                 length:action.arg.length() * sizeof(action.arg[0])
+               encoding:NSUTF16LittleEndianStringEncoding];
+    // SAP-15259 - Add the reply field on a notification
+    if (action.type != electron::NotificationAction::sTYPE_BUTTON)
+      continue;
+    // SAP-14881 - Persistent notification on macOS
+    NSString* actionIdentifier = [[NSString alloc]
+        initWithBytes:action.arg.data()
+               length:action.arg.length() * sizeof(action.arg[0])
+             encoding:NSUTF16LittleEndianStringEncoding];
+    NSUserNotificationAction* notificationAction = [NSUserNotificationAction
+        actionWithIdentifier:actionIdentifier
+                       title:base::SysUTF16ToNSString(action.text)];
+    [additionalActions addObject:notificationAction];
+    additional_action_indices_.insert(
+        std::make_pair(base::SysNSStringToUTF8(actionIdentifier),
+                       additional_action_indices_.size()));
   }
+
   if ([additionalActions count] > 0) {
+    [notification_ setHasActionButton:true];
+    // SAP-14881 - Persistent notification on macOS
+    [notification_ setActionButtonTitle:@"More"];
+    [notification_ setValue:@YES forKey:@"_alwaysShowAlternateActionMenu"];
     [notification_ setAdditionalActions:additionalActions];
   }
 
+  NSMutableDictionary* userInfo = [[NSMutableDictionary alloc] init];
   if (options.has_reply) {
     [notification_ setResponsePlaceholder:base::SysUTF16ToNSString(
                                               options.reply_placeholder)];
     [notification_ setHasReplyButton:true];
+
+    // SAP-15908 - Refine notification-activation for cold start in macOS
+    [userInfo setObject:replyKey_ forKey:@"_replyKey"];
   }
 
+  // SAP-20223 [N] TECH [MACOS] To support renotify on macOS
+  [userInfo setObject:(options.should_be_presented ? @"YES" : @"NO")
+               forKey:@"_shouldBePresented"];
+
   if (!options.close_button_text.empty()) {
     [notification_ setOtherButtonTitle:base::SysUTF16ToNSString(
                                            options.close_button_text)];
   }
 
-  [NSUserNotificationCenter.defaultUserNotificationCenter
-      deliverNotification:notification_];
+  if (!options.data.empty()) {
+    // SAP-14775 - Support data property
+    NSString* dataStr = [[NSString alloc]
+        initWithBytes:options.data.c_str()
+               length:options.data.length() * sizeof(options.data[0])
+             encoding:NSUTF16LittleEndianStringEncoding];
+    // https://developer.apple.com/documentation/foundation/nsusernotification/1415675-userinfo?language=objc
+    // The userInfo content must be of reasonable serialized size (less than
+    // 1KB) or an exception is thrown.
```

Comment:
Is this enforced in the API somewhere to ensure we don't exceed this limit?

Reply:
Not explicitly, I just added warn for Mac but in fact I see that this limit can be supreme on Mac. Also I have feeling that we have bounding into internals of mojom. We need to have additional investigations for the actual value of limit.